### ISSUE TITLE
[FEATURE] Ability to build SOCI indexes when running docker instead of containerd

### ISSUE THREAD

**Issue by bgardner-noggin** [awslabs/soci-snapshotter#805]
### Description

We use bitbucket pipelines to generate and push our images to AWS ECR

We would like to use the SOCI index generator, but this requires containerd to be running which is not possible under bitbucket (it might be possible using custom bitbucket runners though)

An alternative solution was to use the Go 1.x Lambda based generator built by the AWS ECR team https://aws-ia.github.io/cfn-ecr-aws-soci-index-builder/

However, this lambda

1. Is using the Go 1.x runtime which is due to be deprecated in December 2023
2. Builds images which creates invalid "sparse" files. I'm guessing because it is using an older version of the SOCI generator. See https://github.com/aws-ia/cfn-ecr-aws-soci-index-builder/issues/38

### Describe the solution you'd like

The ability for the soci binary to be pull images from local docker

**Comment by Kern--** [comment]
Supporting Docker directly is tricky because Docker doesn't store the compressed image on disk, it compresses during the push:

https://docs.docker.com/engine/reference/commandline/push/
> Progress bars are shown during docker push, which show the uncompressed size. The actual amount of data that's pushed will be compressed before sending, so the uploaded size will not be reflected by the progress bar.

That means we don't have the compressed layers from which to build the SOCI index until after the push. We could try to compress them ourselves before creating the index, but if they're not identical to what Docker eventually pushes, then the indexes wouldn't work (or worse, would behave inconsistently at runtime).

Docker Engine 24.0.0 (released in May 2023) added experimental support for containerd for image management (https://docs.docker.com/storage/containerd/). The SOCI CLI should work with Docker if that flag is enabled, although I haven't had a chance to try it yet. For your use-case with BitBucket, I'm guessing you don't have this level of control over Docker.

There's also the possibility of trying to integrate SOCI index generation into Docker itself. The most likely route for this is [buildkit](https://github.com/moby/buildkit) ([like stargz did](https://github.com/moby/buildkit/blob/master/docs/stargz-estargz.md)), but that still requires configuring buildkit to use containerd under the hood. This isn't out of the question, but I don't think it would directly solve your problems right now.

---

Maybe there's some other way we can unblock your setup in the meantime. I'm not deeply familiar with the hosted bitbucket runners, are you able to run privileged containers? This project's integration tests run on github runners and spin up a privileged container that runs containerd and the SOCI CLI to create indexes. 

https://github.com/awslabs/soci-snapshotter/blob/7cc04aff01bce06f3b1759cf1eecf4cebdeeb2cb/integration/util_test.go#L138

There's also this more complete example of doing the same in AWS CodeBuild: https://github.com/aws-samples/aws-fargate-seekable-oci-toolbox/tree/main/soci-codepipeline.

Would that work for you?

**Comment by bgardner-noggin** [comment]
Running containerd in a container would possibly work. However our usage on SOCI is on hold at the moment as the sparse file bug was found to be a problem in SOCI itself when the --min-layer-size was set to zero - which is what the lambda does. Until that problem is resolved or at least fully understood, we cannot use SOCI in a production environment unfortunately.

**Comment by yob** [comment]
The proposed solution in #1057 would help with this I suspect
